Examples/Io/Performance/ActsExamples/Io/Performance/VertexPerformanceWriter.cpp

```diff
@@ -243,219 +242,175 @@ ActsExamples::ProcessCode ActsExamples::VertexPerformanceWriter::writeT(
-            contributingTruthVertices.push_back(vtxId);
-            foundMatchingParams = true;
+        // parameters. This allows us to identify the corresponding particle.
+        // TODO this should not be necessary if the tracks at vertex would keep
+        // this information
+        for (const auto& inputTrk : *tracks) {
+          const auto& params = inputTrk.parameters();
+
+          if (origTrack.parameters() == params) {
+            trackIndices.push_back(inputTrk.index());
+            foundMatchingParticle = true;
+
+            if (trk.trackWeight > m_cfg.minTrkWeight) {
+              auto particleOpt = findParticle(inputTrk);
+              if (!particleOpt.has_value()) {
+                continue;
+              }
+              const auto& particle = *particleOpt;
+
+              particleAtVtx.insert(particle);
+              SimBarcode vtxId = SimBarcode(particle.particleId())
+                                     .setParticle(0)
+                                     .setSubParticle(0);
+              contributingTruthVertices.insert(vtxId);
+            }
+
             break;
           }
         }
-        if (!foundMatchingParams) {
+
+        if (!foundMatchingParticle) {
           ACTS_VERBOSE("Track has no matching truth particle.");
         }
-      }  // end loop tracksAtVtx
+      }
+
+      if (tracksAtVtx.size() != trackIndices.size()) {
+        ACTS_ERROR("Not all tracks at vertex have a matching input track.");
+      }
     } else {
-      for (const auto& particle : allTruthParticles) {
+      for (const auto& particle : particles) {
         SimBarcode vtxId =
-            particle.particleId().setParticle(0).setSubParticle(0);
-        contributingTruthVertices.push_back(vtxId);
+            SimBarcode(particle.particleId()).setParticle(0).setSubParticle(0);
+        contributingTruthVertices.insert(vtxId);
       }
     }
 
     // Find true vertex that contributes most to the reconstructed vertex
-    std::map<SimBarcode, int> fmap;
-    for (const SimBarcode& vtxId : contributingTruthVertices) {
+    std::map<SimBarcode, std::uint32_t> fmap;
+    for (SimBarcode vtxId : contributingTruthVertices) {
       fmap[vtxId]++;
     }
-    int maxOccurrence = -1;
-    SimBarcode maxOccurrenceId = -1;
-    for (auto it : fmap) {
-      if (it.second > maxOccurrence) {
-        maxOccurrenceId = it.first;
-        maxOccurrence = it.second;
+    std::uint32_t maxOccurrence = 0;
+    SimBarcode maxOccurrenceId;
+    for (const auto& [vtxId, count] : fmap) {
+      if (count > maxOccurrence) {
+        maxOccurrenceId = vtxId;
+        maxOccurrence = count;
       }
     }
 
     // Count number of reconstructible tracks on truth vertex
-    int nTracksOnTruthVertex = 0;
-    for (const auto& particle : associatedTruthParticles) {
-      SimBarcode vtxId = particle.particleId().setParticle(0).setSubParticle(0);
+    std::uint32_t nTracksOnTruthVertex = 0;
+    for (const auto& particle : recoParticles) {
```

Comment:
Isn't this confusing "resonstructible" and "reconstructed" particles? Previously, the assumption was (I think) that `associatedTruthParticles` where all reconstructible, but now it's derived from the actual calculated tracks.

Reply:
I think even before `associatedTruthParticles` was doing "reconstructed" particles. But you are right this should rather be "reconstructible" particles.